### ISSUE TITLE
Prometheus: More detailed metrics

### ISSUE THREAD

**Issue by danielh1989** [containous/traefik#3196]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?
I like to have a more detailed metric with prometeus.
For example:

- Requests per Route
- Requests per Backend
- Average Time per Backend
- Average Time per Route

**Comment by mmatur** [comment]
Hi @danielh1989,

Thanks for your interest in Træfik projet.

For information, the current available metrics are :

- config_reloads_total
- traefik_config_reloads_failure_total
- traefik_config_last_reload_success
- traefik_config_last_reload_failure
- traefik_entrypoint_requests_total
- traefik_entrypoint_request_duration_seconds
- traefik_entrypoint_open_connections
- traefik_backend_requests_total
- traefik_backend_request_duration_seconds
- traefik_backend_open_connections
- traefik_backend_retries_total
- traefik_backend_server_up

**Comment by marco-jantke** [comment]
What do you mean exactly when you tell "per route"? Prometheus doesn't like "unbound" label values and so putting an arbitrary path into it is not an option if you don't want to kill your Prometheus server.

**Comment by Place1** [comment]
as an example, i'd really like to know which API (based on a path) is slower than others when using the traefik_backend_requests_total metric. my services don't have an unbounded number of APIs so perhaps Traefik could only fill the label for backend requests that are not a 404 to avoid client spam.

**Comment by mgerasimchuk** [comment]
Recently I moved from ingress-nginx controller to traefik, and only now I realised, that traefik unfortunately doesn't have any way to collect metrics by URL path. For example in ingress-nginx controller this behavior works by default: https://github.com/kubernetes/ingress-nginx/blob/96b6228a6b65a85e421b8a348a149e99181664d1/deploy/grafana/dashboards/request-handling-performance.json#L314

Traefik has a lot of pros like SSL certificates management and very comfortable routing, but observability of the system also is one of the most important thing, and now I can't understand which route of my API works slow....
Hope that the `path` label will be added in nearest future..

**Comment by rayjanoka** [comment]
I found the metric labels were more usable if I used an intermediate `TraefikService` in the path to my actual k8s service.

**Comment by hick** [comment]
> Recently I moved from ingress-nginx controller to traefik, and only now I realised, that traefik unfortunately doesn't have any way to collect metrics by URL path. For example in ingress-nginx controller this behavior works by default: https://github.com/kubernetes/ingress-nginx/blob/96b6228a6b65a85e421b8a348a149e99181664d1/deploy/grafana/dashboards/request-handling-performance.json#L314
> 
> Traefik has a lot of pros like SSL certificates management and very comfortable routing, but observability of the system also is one of the most important thing, and now I can't understand which route of my API works slow.... Hope that the `path` label will be added to metrics in nearest future..

yes, the observability for every path of URL(route) is important for me too,  even for alert/notification of traffic accident.

**Comment by lol768** [comment]
Needing to add a new header to get this to work is a really ugly hack. C'mon, this should be pretty basic functionality available out of the box (and it _is_ available out of the box w/ ingress-nginx, with ASP.NET Core's [built-in metrics](https://learn.microsoft.com/en-us/aspnet/core/log-mon/metrics/metrics?view=aspnetcore-9.0) etc etc). All you really need is something in the docs to say "make damn sure you group your paths and understand their boundedness or you'll make your Prom server sad".

I really do reject this notion that it suddenly becomes a breaking change _in the software_ because a user becomes able to do something dumb with it and cause themselves pain. If you play stupid games, you win stupid prizes. @Place1's suggestion for stripping out 404s is a sensible one; then you just need to have an understanding of the paths your routing framework will accept.

**Comment by rtribotte** [comment]
Hello,

Thanks for all the feedback!
However, there is still something unclear to me @hick @lol768 @mgerasimchuk @Place1 

Do you want the request path or the router's rule path to be added as a label to metrics?

**Comment by mgerasimchuk** [comment]
Hi @rtribottem,

Ideally, we want to have an HTTP Request Path.

But at least for me, it would be also a good alternative if we would be able to have at least Traefik route Rule

**Comment by jkblume** [comment]
Hello, I raise my hand for this feature, especially for the information about the request path. I do not fully understand the problem with the route rule or http request path. IMHO the http request path should be used which is the current one passed to the metrics middleware.
